flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -336,6 +336,8 @@ public CompletedCheckpoint finalizeCheckpoint(
                                 props,
                                 finalizedLocation);
 
+                // Mark this pending checkpoint as disposed, but do NOT drop the state.
+                dispose(false, checkpointsCleaner, postCleanup, executor);
```

Comment:
Can you explain how exactly the race condition happens and how exactly moving `dispose` before `onCompletionPromise` fixes it?

Reply:
Sure:

- During scheduler cleanup, checkpoint cleaner gets closed right after all of the checkpoint futures complete [1]
- CheckpointsCleaner close future completes when all of the async disposals are finished (all data are removed). This uses a simple counter (`numberOfCheckpointsToClean`) to track pending disposals.
- We're not allowed to register new disposals after CheckpointsCleaner#closeAsync has been called
- Calling #dispose() tries to register a new disposal with CheckpointsCleaner [2]

-> We need to register all disposals before closing the checkpointsCleaner

[1] https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java#L606
[2] https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java#L106

Reply:
Shouldn't we make sure that nothing uses/adds new things to clean up to the `CheckpointsCleaner` before shutting it down? It seems like a brittle contract that we need to place the `dispose` call at certain places in the code.

Reply:
How do you want to ensure this? We already fail any call to `#cleanupCheckpoint` / `#cleanCheckpointOnFailedStoring` after CheckpointsCleaner has been closed.

This `dispose()` method you're referring to is private to the PendingCheckpoint implementation. PendingCheckpoint implementation is the one that needs to say, "I'm done with the checkpoint and you can treat it as completed." by completing the checkpoint future. If this class then registers checkpoint for cleanup (which is basically mutation of the result it has already returned), then it's IMO a problem with the implementation and not the CheckpointsCleaner contract.

WDYT?

Reply:
I think the problem is in the `SchedulerBase` and the `AdaptiveScheduler` how they shut down the `CheckpointsCleaner`. If we stop the `CheckpointsCleaner` after the `ExecutionGraph` has terminated and all completed checkpoints have been released, then this problem should hopefully be gone. 

The underlying problem is that the `ExecutionGraph` is using a service that stops working in the midst of some method (`finalizeCheckpoint`). This should not happen since this is super hard to reason about.

Reply:
Aaaaaaa, I've finally got the missing piece 🤦  🤦 

```java
        final CompletableFuture<Void> checkpointServicesShutdownFuture =
                CompletableFuture.allOf(
                        executionGraph
                                .getTerminationFuture()
                                .thenAcceptAsync(
                                        this::shutDownCheckpointServices, getMainThreadExecutor()),
                        checkpointsCleaner.closeAsync());
```

I've read this statement as "close the cleaner after the execution graph" for the whole time

Now what you're suggesting finally starts making sense :)

Reply:
Fixed. I've removed the old patch as it's not necessary anymore.